File: fenwick-tree/src/FenwickTree.java
public class FenwickTree {

    private int n;
    private int[] tree;

    public int[] getTree() {
        return tree;
    }

    public FenwickTree(final int n) {
        this.n = n;
        this.tree = new int[this.n + 1];
    }

    public FenwickTree(final int[] arr) {
        this.n = arr.length;
        this.tree = new int[arr.length + 1];

        for (int i = 0; i < n; i++) {
            update(i + 1, arr[i]);
        }
    }

    public void update(int i, int value) {
        while (i <= this.n) {
            tree[i] += value;
            System.out.println("Position on update:" + (i & -i));
            i += i & -i;
        }
    }

    // Not working
    public void create(int[] arr) {
        for (int i = 0; i < arr.length; i++) {

            int y = i;
            while (y < n) {
                tree[y] += arr[i];
                i += i & -i;
            }
        }
    }

    public int query(int i) {
        int sum = 0;
        while (i > 0) {
            sum += tree[i];
            System.out.println("Position on query:" + (i & -i));
            i -= i & -i;
        }
        return sum;
    }

    public int queryRange(int p1, int p2) {
        return query(p2) - query(p1 - 1);
    }

}

File: fenwick-tree/kotlin/src/FenwickTree.kt
class FenwickTree {

    private var n: Int
    private var tree: IntArray

    constructor(n: Int) {
        this.n = n
        this.tree = IntArray(n + 1)
    }

    constructor(arr: IntArray) {
        this.n = arr.size
        this.tree = IntArray(arr.size + 1)
        for (i in arr.indices) {
            update(i + 1, arr[i])
        }
    }

    fun getTree(): IntArray = tree

    fun update(index: Int, value: Int) {
        var i = index
        while (i <= n) {
            tree[i] += value
            i += i and -i
        }
    }

    fun sum(index: Int): Int {
        var i = index
        var sum = 0

        while (i > 0) {
            sum += tree[i]
            i -= i and -i
        }

        return sum
    }

    fun sumBetween(startIndex: Int, endIndex: Int): Int {
        return sum(endIndex) - sum(startIndex - 1)
    }
}
